File: python/receive.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Aug 28 12:03:30 2019

@author: drone
"""

import socket
import threading
import time
import requests
import json
import os
import sys

host = ""
port = 9000
data_dir = sys.argv[1]
api_url = sys.argv[2]

picture_extension = ['.jpg', '.png', 'jpeg']
video_extension = ['.mov', '.avi', '.mp4']
id_socket = {"mission" : "", "intervention" : ""}
serversock = socket.socket()
serversock.bind((host, port))
serversock.listen(20)
sys.stdout.write("Waiting for files...")
sys.stdout.flush()

def save(client_socket, collection):
    global data_dir
    size = client_socket.recv(16)
    size = int(size, 2)
    file_name = client_socket.recv(size)
    file_size = client_socket.recv(32)
    file_size = int(file_size, 2)
    file_path = data_dir+file_name.decode()
    file_to_write = open(file_path, 'wb')
    chunk_size = 4096
    while file_size > 0:
        if file_size < chunk_size:
            chunk_size = file_size
        data = client_socket.recv(chunk_size)
        file_to_write.write(data)
        file_size -= len(data)
    file_to_write.close()
    time.sleep(0.001)
    f_inter = open(file_path, 'r')
    address = api_url+collection+"/"
    content_json = json.loads(f_inter.read())
    requests.post(address, json = content_json)
    id_socket[collection] = json.loads(requests.get(address, json = content_json).text)[0]['_id']
    new_dir = data_dir+collection+id_socket[collection]+"/"
    if not os.path.exists(new_dir):
        os.system("mkdir "+new_dir)
    else:
        os.system("rm "+str(new_dir+file_name.decode()))
    os.system("mv "+str(file_path)+" "+str(new_dir+file_name.decode()))
    data_dir = new_dir

def reader(client_socket):
    save(client_socket, "intervention")
    save(client_socket, "mission")
    requests.post(api_url+"link_mission_intervention/",
              json = {'intervention_id':str(id_socket["intervention"]), 
                      'mission_id':str(id_socket["mission"])})
    while True:
        size = client_socket.recv(16)
        if not size:
            break
        size = int(size, 2)
        file_name = client_socket.recv(size).decode()
        file_extension = file_name[len(file_name)-4:len(file_name)]
        file_size = client_socket.recv(32)
        file_size = int(file_size, 2)
        file_path = data_dir+file_name
        file_to_write = open(file_path, 'wb')
        chunk_size = 4096
        while file_size > 0:
            if file_size < chunk_size:
                chunk_size = file_size
            data = client_socket.recv(chunk_size)
            file_to_write.write(data)
            file_size -= len(data)
        file_to_write.close()
        sys.stdout.write(file_name+' received successfully')
        sys.stdout.flush()
        if file_extension in picture_extension:
            requests.post(api_url+"frame/", json = {"frame_path" : file_path})
            picture_id = json.loads(requests.get(api_url+"frame/", 
                                                 json = {"frame_path" : file_path}).text)[0]['_id']
            requests.post(api_url+"link_frame_mission/", json = {'frame_id':str(picture_id), 
                      'mission_id':str(id_socket["mission"])})
        elif file_extension in video_extension:
            video_json = {"video_path" : file_path, "video_extension" : file_extension}
            requests.post(api_url+"video/", json = video_json)
            video_id = json.loads(requests.get(api_url+"video/", json = video_json).text)[0]['_id']
            requests.post(api_url+"link_video_mission/", json = {'video_id':str(video_id), 
                      'mission_id':str(id_socket["mission"])})

if __name__ == "__main__":
    while True:
        client, addr = serversock.accept()
        sys.stdout.write("Got a connection from %s" % str(addr))
        sys.stdout.flush()
        client_serve_thread = threading.Thread(target=reader, args=tuple((client,)))
        client_serve_thread.start()
        time.sleep(0.001)
    serversock.close()

File: fileManager/receive.js
/**
 * http://usejsdoc.org/
 */

const io = require("socket.io");
const fs = require('fs');
const request = require('request');

const server = io.listen(3000);
const location =  '/home/drone/data/';
const api_url = 'http://127.0.0.1:1337/';

const video_extension = ['.mov', '.avi', '.mp4'];
const picture_extension = ['.jpg', '.png', 'jpeg'];

var link_saved;
var mission_saved;
var intervention_saved;
var address;
var extension;
var link_mission_intervention;
var link_frame_mission;
var link_video_mission;
var frame;
var video;

server.on("connection", async function(socket) {
	mission_saved = false;
	intervention_saved = false;
	link_saved = false;
	link_mission_intervention = new Object();
	link_frame_mission = new Object();
	link_video_mission = new Object();
	console.log("New connection.");
	socket.emit('msg', 'Connected!');
	socket.on('sendFile', function(data, filename){
		var buff = Buffer.from(data, 'base64');
		address = location+filename;
		fs.writeFileSync(address, buff);
		extension = filename.substr(filename.length-4);
		// save the intervention
		if(filename === 'intervention.json'){
			fs.readFile(address, 'utf8', function(err, jsonString){
			    if (err) {
			        console.log("intervention.json impossible to read :", err)
			        return
			    }
			    //console.log(JSON.parse(jsonString));
			    request.post(api_url+'intervention/', {json: JSON.parse(jsonString)}, 
			    	  (error, res, body) => {
				    	  if (error) {
				    	    console.error(error);
				    	    return
				    	  }
			    	  //console.log(`statusCode: ${res.statusCode}`);
				      link_mission_intervention.intervention_id = body['_id'];
			    });
			})
			intervention_saved = true;
			socket.emit('msg', filename+' received from client and saved!');
		}
		//save the mission
		if(filename === 'mission.json'){
			fs.readFile(address, 'utf8', function(err, jsonString){
			    if (err) {
			        console.log("mission.json impossible to read :", err);
			        return
			    }
			    //console.log('File data:', JSON.parse(jsonString));
			    request.post(api_url+'mission/', {json: JSON.parse(jsonString)}, 
			    	  function(error, res, body){
				    	  if (error) {
				    	    console.error(error);
				    	    return
				    	  }
			    	  //console.log(`statusCode: ${res.statusCode}`);
				      link_mission_intervention.mission_id = body['_id'];
				      link_video_mission.mission_id = body['_id'];
				      link_frame_mission.mission_id = body['_id'];
			    });
			})
			mission_saved = true;
			socket.emit('msg', filename+' received from client and saved!');
		}
		// once you saved the intervention and the mission, save the link between intervention and mission
		if((link_mission_intervention.mission_id !== undefined) && (link_mission_intervention.intervention_id !== undefined) && !link_saved){
			request.post(api_url+'link_mission_intervention/', {json: link_mission_intervention}, 
			    	  function(error, res, body){
				    	  if (error) {
				    	    return
				    	  }
			    	  console.log('Link between intervention and mission ok!');
			    	  link_saved = true;
			});
		}
		if(picture_extension.indexOf(extension)>=0){	
			frame = new Object();
			frame.frame_path = location+filename;
			request.post(api_url+'frame/', {json: frame}, 
			    	  function(error, res, body){
				    	  if (error) {
				    	    return
				    	  }
			    	  link_frame_mission.frame_id = body['_id'];
			    	  link_frame_mission.mission_id = link_mission_intervention.mission_id;
			    	  request.post(api_url+'link_frame_mission/', {json: link_frame_mission}, 
					    	  function(error, res, body){
						    	  if (error) {
						    	    return
						    	  }
						      console.log('Link between frame and mission ok!');
			    	  });
			});
			console.log('File received from client! Picture saved at : '+location+filename);
			socket.emit('msg', 'Picture received from client! '+filename+' saved!');
		}
		if(video_extension.indexOf(extension)>=0){
			video = new Object();
			video.video_path = location+filename;
			video.video_extension = extension;
			if(link_video_mission.mission_id !== undefined){
				request.post(api_url+'video/', {json: video}, 
			    	  function(error, res, body){
				    	  if (error) {
				    	    return
				    	  }
			    	  link_video_mission.video_id = body['_id'];
			    	  link_video_mission.mission_id = link_mission_intervention.mission_id;
			    	  console.log(link_video_mission);
			    	  request.post(api_url+'link_video_mission/', {json: link_video_mission}, 
					    	  function(error, res, body){
						    	  if (error) {
						    	    return
						    	  }
						      console.log('Link between video and mission ok!');
			    	  });
				});
			}
			console.log('File received from client! Video saved at : '+location+filename);
			socket.emit('msg', 'Video received from client! '+filename+' saved!');
		}
	})
});
